feat: add more spec to transaction processing and errors (#108)
nearprotocol/NEPs: specs/RuntimeSpec/Actions.md

- sets Account `storage_usage` to `account_cost` (genesis config)

> - sets Account `storage_paid_at` to the current block height

Reviewer: This is no more

**Errors**:
- If the action tries to create a top level account whose length is no greater than 32 characters, and `predecessor_id` is not
`registrar_account_id`, the following error will be returned
```rust
/// A top-level account ID can only be created by registrar.
CreateAccountOnlyByRegistrar {
    account_id: AccountId,
    registrar_account_id: AccountId,
    predecessor_id: AccountId,
}
```

- If the action tries to create an account that is neither a top-level account or a subaccount of `predecessor_id`,
the following error will be returned
```rust
/// A newly created account must be under a namespace of the creator account
CreateAccountNotAllowed { account_id: AccountId, predecessor_id: AccountId },
```

> NOTE: for the all subsequent actions in the transaction the `signer_id` becomes `receiver_id` until [DeleteAccountAction](#DeleteAccountAction). It allows to execute actions on behalf of the just created account.

Reviewer: Found this note. good. See comment below for `actor_id`

Instead of `signer_id` you should use `predecessor_id`. Actions can be created through a promise API without extra transaction. Since this section talks about Actions in general context and not only actions within a transaction, then we should use `predecessor_id` whenever possible and forget about `signer_id`

## DeployContractAction

```rust
pub struct DeployContractAction {
    pub code: Vec<u8>, // a valid WebAssembly code
}
```

_Requirements:_

> - _`tx.signer_id` to be equal to `receiver_id`_

Reviewer: In runtime we used a concept of an actor_id. This is the temporary account ID that is either the `predecessor_id` (not `signer_id`) or the current account ID. By default it's the predecessor, but if the first action is `CreateAccount` and the account doesn't exist, it changes the actor to current account ID `receiver_id`

- _`tx.public_key` to be `AccessKeyPermission::FullAccess` for the `signer_id`_

**Outcome**:
- sets the contract code for account

**Errors**

> - If state or storage is corrupted, it may return `StorageError`. Otherwise this action should not lead to errors.

Reviewer: I don't think spec should talk about storage errors. Until we get to challenges, you shouldn't be able to arrive into inconsistent state

Author: I think it is part of the spec because it is possible that we return `StorageError` if state is somehow corrupted.

}
```

_Requirements:_

- _`tx.public_key` to be `AccessKeyPermission::FullAccess` for the `singer_id`_

**Outcome**:
- transfers amount specified in `deposit` from `tx.signer` to a `tx.receiver_id` account

**Errors**:
- If the deposit amount plus the existing amount on the receiver account exceeds `u128::MAX`,

> a `StorageInconsistentState("Account balance integer overflow")` error will be returned.

Reviewer: I think you can take a look at `verifiers.rs` and tests to see what errors are thrown during transaction validation.

There are some balance overflow, but it's not inconsistent state error.

There are also execution time errors, such as `AccountExists` that are thrown in `actions.rs`

Author: https://github.com/nearprotocol/nearcore/blob/1c1c019613e4ccb47ea6dc5a591f913af8022998/runtime/runtime/src/actions.rs#L293

Reviewer: That's mostly only possible for challenges. We wrapped most errors in order to make sure when the challenge on malicious state transition arrives (where the previous root is also malicious), we don't crash the node.

    pub public_key: PublicKey,
}
```

**Outcome**:
- A validator proposal that contains the staking public key and the staking amount is generated and will be included
in the next block.

### Errors

**Validation Error**:

> - If the `public_key` is not an ed25519 key, the following error will be returned:

Reviewer: ```suggestion
- If the `public_key` is not an ristretto compatible ed25519 key, the following error will be returned:
```

}
```

**Outcome**:

> - Associates an [AccessKey](AccessKey) with a `public_key` provided.

Reviewer: It also adds this AccessKey to the receiver's account. 
```suggestion
- Adds a new [AccessKey](AccessKey) to the receiver's account and associates it with a `public_key` provided.
```

### Errors:

**Validation Error**:

If the access key is of type `FunctionCallPermission`, the following errors can happen
- If `receiver_id` in `access_key` is not a valid account id, the following error will be returned 
```rust
/// Invalid account ID.
InvalidAccountId { account_id: AccountId },
```

- If the length of some method name exceed `max_length_method_name`, which is a genesis parameter (current value is 256),
the following error will be returned 
```rust
/// The length of some method name exceeded the limit in a Add Key action.
AddKeyMethodNameLengthExceeded { length: u64, limit: u64 },
```

> - If the sum of length of method names exceeds `max_number_bytes_method_names`, which is a genesis parameter (current value is 2000),

Reviewer: ```suggestion
- If the sum of length of method names (with 1 extra character for every method name) exceeds `max_number_bytes_method_names`, which is a genesis parameter (current value is 2000),
```

the following error will be returned
```rust
/// The total number of bytes of the method names exceeded the limit in a Add Key action.
AddKeyMethodNamesNumberOfBytesExceeded { total_number_of_bytes: u64, limit: u64 }
```

**Execution Error**:

> - If an account tries to add an access key that already exists, the following error will be returned

Reviewer: ```suggestion
- If an account tries to add an access key with a given public key, but an existing access key with this public key already exists, the following error will be returned
```

    pub public_key: PublicKey,
}
```

**Outcome**:
- Deletes the [AccessKey](AccessKey) associated with `public_key`.

### Errors

**Execution Error**:

- When an account tries to delete an access key that doesn't exist, the following error is returned
```rust
/// Account tries to remove an access key that doesn't exist
DeleteKeyDoesNotExist { account_id: AccountId, public_key: PublicKey }
```
- `StorageError` is returned if state or storage is corrupted.

## DeleteAccountAction

```rust
pub struct DeleteAccountAction {
    /// The remaining account balance will be transferred to the AccountId below
    pub beneficiary_id: AccountId,
}
```

**Outcomes**:
- The account, as well as all the data stored under the account, is deleted and the tokens are transferred to `beneficiary_id`.

### Errors

**Validation Error**
- If `beneficiary_id` is not a valid account id, the following error will be returned
```rust
/// Invalid account ID.
InvalidAccountId { account_id: AccountId },
```

- If this action is not the last action in the action list of a receipt, the following error will be returned
```rust
/// The delete action must be a final action in transaction
DeleteActionMustBeFinal
```

> - If the account still has locked balance, the following error will be returned

Reviewer: ```suggestion
- If the account still has locked balance due to staking, the following error will be returned
```
